lib/constructs/interface.js

```diff
@@ -1644,6 +1667,73 @@ class Interface {
     `;
   }
 
+  generateLegacyFactoryFunction() {
+    const { legacyFactoryFunctions } = this;
+    if (legacyFactoryFunctions.length === 0) {
+      return;
+    }
+
+    const name = legacyFactoryFunctions[0].rhs.value;
+
+    if (!name) {
+      throw new Error(`Internal error: The legacy factory function does not have a name (in interface ${this.name})`);
+    }
+
+    const overloads = Overloads.getEffectiveOverloads("legacy factory function", name, 0, this);
+    let minOp = overloads[0];
+    for (let i = 1; i < overloads.length; ++i) {
+      if (overloads[i].nameList.length < minOp.nameList.length) {
+        minOp = overloads[i];
+      }
+    }
+
+    const args = minOp.nameList;
+    const conversions = Parameters.generateOverloadConversions(
+      this.ctx,
+      "legacy factory function",
+      name,
+      this,
+      `Failed to construct '${name}': `
+    );
+    this.requires.merge(conversions.requires);
+
+    const argsParam = conversions.hasArgs ? "args" : "[]";
+
+    this.str += `
+      function ${name}(${utils.formatArgs(args)}) {
+        if (new.target === undefined) {
+          throw new TypeError("Class constructor ${name} cannot be invoked without 'new'");
+        }
+
+        ${conversions.body}
+    `;
+
+    // This implements the WebIDL legacy factory function behavior, as well as support for overridding
+    // the return type, which is used by HTML's element legacy factory functions:
```

Comment:
This seems more likely to be a historical artifact of how the spec is written?

It seems like we should be able to support either return overload (like HTML uses for its LegacyFactoryFunctions), or generate a this value (like modern Web IDL-using constructor specs do). We shouldn't need both, I don't think?

In particular given that you can subclass `Option` per the tests you wrote, I suspect the first line of https://html.spec.whatwg.org/#dom-option is just incorrect.

Reply:
It occurred to me that the order of `Impl.legacyFactoryFunction` and `Impl.init` calls is wrong, so I’m fixing that so that the object returned from `Impl.legacyFactoryFunction` is now the `impl`, and the result of `makeWrapper` is what’s returned from the `LegacyFactoryFunction` constructor.